Enhancing tagging interface

zachmprince opened the pull request:
Closes #23785 #23787

zachmprince committed:
Adding component index to Coupleable::coupledVectorTagArrayDofValue #23785

zachmprince committed:
Including tagged vectors in ComputeNodalUserObjectsTread #23785

zachmprince committed:
Accumulating tagged vectors in EigenKernel #23785

zachmprince committed:
Only accumulating tagged vectors that are requested #23787

lindsayad commented:
is this ready for review?

zachmprince commented:
> is this ready for review?

@lindsayad Actually I found another issue. I'm going to create a test. I'll mark as draft until it's ready.

lindsayad commented:
alright in the case that github doesn't notify me when it's ready for review, mention me

zachmprince committed:
Adding test to make sure tagging changes work properly #23787

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1414701/) on 61988e0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23788/site)

This comment will be updated on new commits.

lindsayad committed:
Update some missed residual adding/caching APIs

lindsayad committed:
Restore Problem cacheResidual APIs

lindsayad committed:
Create _current_residual_vector_tags

This is set by calls to `FEProblemBase::computeResidualTags` and
`FEProblemBase::computeResidualAndJacobian`. These tags get sent
to `Assembly` when caching and adding residuals

Refs #23785 #23787

lindsayad committed:
Second round of restoring residual APIs

lindsayad committed:
Complete switch to use currentResidualVectorTags

Refs #23785 #23787

lindsayad committed:
Set the current residual vector tag from CrankNicolson

And protect NonlinearSystemBase::computeResidualTag(s) calls from
being made outside of the framework library. Similarly for
the FEProblemBase current residual vector tag controlling methods

lindsayad committed:
Don't protect tag computation APIs with Passkey

The rooster has already flown the coup with Griffin using
`NonlinearSystemBase::computeResidualTag`

zachmprince commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5859,6 +5856,8 @@ FEProblemBase::computeResidualAndJacobian(const NumericVector<Number> & soln,
     for (const auto & residual_vector_tag : residual_vector_tags)
       _fe_vector_tags.insert(residual_vector_tag._id);
+
+    setCurrentResidualVectorTags(_fe_vector_tags);
```
I'm a little confused on why this needs to be FEProblemBase and can't be pushed down to NonlinearSystemBase. Is it because `setCurrentResidualVectorTags` is telling what vectors to accumulate and the `tags` function parameter is telling what residual objects to go through?

lindsayad replied:
To me it's about data locality. In almost all cases the tag computation begins with `_fe_problem.computeResidualTags`. And then when it comes time to accumulate residuals into the global data structures it happens through  `_fe_problem.XXXResidual`. Any time I call `_fe_problem.XXXResidual` the problem would have to query the nonlinear system base object for the residual vector tags; I just don't see the reason for the indirection

lindsayad replied:
Additionally, the problem objects are where are all the other vector tag data is stored, e.g. `_vector_tags`, `_typed_vector_tags`, etc.

zachmprince replied:
Ahh, that makes more sense. I just thought it strange that CrankNicolson had to call this function too.

lindsayad replied:
What's unfortunate is that CrankNicolson has to call the `nl.computeResidualTag` function directly. This direct call also happens in some Griffin solve objects. I would have liked to make that function private if I could have

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1423011/) on 1e4e25b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8f6c66</th>
      <th colspan=4>#23788 1e4e25</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.89%</td>
      <td>84.89%</td>
      <td>-0.00%</td>
      <td>77.98%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84301</td>
      <td>84333</td>
      <td>+32</td>
      <td>85</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15003</td>
      <td>15014</td>
      <td>+11</td>
      <td>24</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23788/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23788/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23788/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 77.98% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented:
@zachmprince can you add some testing of the new lines you added in `EigenKernel`?

zachmprince commented:
> @zachmprince can you add some testing of the new lines you added in `EigenKernel`?

Sure I'll add some tests.

zachmprince committed:
Adding tests for EigenKernel tag accumulation #23785

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1427205/) on 8146a3c : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1427223/) on 8146a3c wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 86.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 83.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 55.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 89.0%)
- Failed to generate `porous_flow` coverage rate (required: 96.0%)
- Failed to generate `ray_tracing` coverage rate (required: 94.0%)
- Failed to generate `rdg` coverage rate (required: 65.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 84.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 86.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1427223/) on 8146a3c wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 86.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 83.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 55.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 89.0%)
- Failed to generate `porous_flow` coverage rate (required: 96.0%)
- Failed to generate `ray_tracing` coverage rate (required: 94.0%)
- Failed to generate `rdg` coverage rate (required: 65.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 84.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 86.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

loganharbour commented:
Why did you force push?

lindsayad commented:
To get a passing coverage target

lindsayad commented:
nothing else changed other than rebasing on "current" devel

loganharbour commented:
If you just invalidate all, it'll do that

lindsayad commented:
But I think our previous branch base commit may not have been a commit that had coverage generated for it. Maybe it had, I dunno. I'll try invalidating all next time if I don't know for sure

loganharbour commented:
This broke next: https://civet.inl.gov/event/117524/

lindsayad commented:
Nice. I'll look into it soon unless @zachmprince is already doing so
